libmscore/cmd.cpp

```diff
@@ -204,41 +204,67 @@ void CmdState::dump()
       }
 #endif
 
+//---------------------------------------------------------
+//   update
+//    layout & update
+//---------------------------------------------------------
+Score* Score::scoreOfUpdate = 0;
+
+void Score::doConcurrentLayoutAndUpdate()
+      {
+      CmdState& cs = _masterScore->cmdState();
+
+      if (cs.layoutRange())
+            doLayoutRange(cs.startTick(), cs.endTick());
+
+      if (cs.updateAll()) {
+            for (MuseScoreView* v : viewer)
+                  v->updateAll();
+            }
+      else if (cs.updateRange() && this == Score::scoreOfUpdate) {
+            // updateRange updates only current score
+            qreal d = spatium() * .5;
+            _updateState.refresh.adjust(-d, -d, 2 * d, 2 * d);
+            for (MuseScoreView* v : viewer)
+                  v->dataChanged(_updateState.refresh);
+            _updateState.refresh = QRectF();
+            }
+      }
+
 //---------------------------------------------------------
 //   update
 //    layout & update
 //---------------------------------------------------------
 
 void Score::update()
```

Comment:
(as an aside, in my opinion, the name of the main function "Score::update()" is a little confusing, because it actually acts upon all scores in the MasterScore, and it does both layout (if necessary) and repaint (if necessary)...I think there is a better name for this, but I don't think I have permission to change it.)